cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -169,26 +182,32 @@ MapBuilderBridge::GetTrajectoryStates() {
     const int trajectory_id = entry.first;
     const SensorBridge& sensor_bridge = *entry.second;
 
-    const cartographer::mapping::TrajectoryBuilder* const trajectory_builder =
-        map_builder_.GetTrajectoryBuilder(trajectory_id);
-    if (trajectory_builder == nullptr) {
-      continue;
+    std::unique_ptr<TrajectoryState::LocalSlamData>
+        current_trajectory_state_data;
+    {
+      cartographer::common::MutexLocker lock(&mutex_);
+      if(trajectory_state_data_.count(trajectory_id) == 0) {
+        continue;
+      }
+      current_trajectory_state_data =
```

Comment:
`GetTrajectoryStates()` gets called at a configurable (and usually high) frequency. Even if we decide not to publish the point cloud again, we copy it here. Did you consider making this a `shared_ptr` to avoid the copy?

Reply:
Yeah, I did, but in libcartographer, [if you recall](https://github.com/googlecartographer/cartographer/pull/574#discussion_r147361926). We removed it during the review and it didn't occur to me to reintroduce it here.

Sure, I will update it to use a shared pointer.

Reply:
Not a noticeable regression. I haven't measured though. I updated it to use shared pointers. A bonus is that assigning shared pointers should be thread safe, so I also got rid of the mutex.

Reply:
No, that is not correct! That is a common fallacy. See https://youtu.be/lkgszkPnV8g?t=1213 for a decent discussion of this issue. 

You have to take a lock or use atomic operations on your shared_ptr. Please do not use atomic operations on your shared pointer - they are subtle and never do what you think they do. Instead take take a lock.